### PR TITLE
[Clang][Comments] Attach comments to decl even if preproc directives are in between

### PR DESCRIPTION
**Pull request by hdoc**
### Background

It's surprisingly common for C++ code in the wild to conditionally show/hide declarations to Doxygen through the use of preprocessor directives. One especially common version of this pattern is demonstrated below:

```cpp
/// @brief Test comment
#ifdef DOXYGEN_BUILD_ENABLED
template<typename T>
#else
template <typename T>
typename std::enable_if<std::is_integral<T>::value>::type
#endif
void f() {}
```

There are more examples I've collected below to demonstrate usage of this pattern:
- Example 1: [Magnum](https://github.com/mosra/magnum/blob/8538610fa27e1db37070eaabe34f1e4e41648bab/src/Magnum/Resource.h#L117-L127)
- Example 2: [libcds](https://github.com/khizmax/libcds/blob/9985d2a87feaa3e92532e28f4ab762a82855a49c/cds/container/michael_list_nogc.h#L36-L54)
- Example 3: [rocPRIM](https://github.com/ROCm/rocPRIM/blob/609ae19565ff6a3499168b76a0be5652762e24f6/rocprim/include/rocprim/block/detail/block_reduce_raking_reduce.hpp#L60-L65)
 
From my research, it seems like the most common rationale for this functionality is hiding difficult-to-parse code from Doxygen, especially where template metaprogramming is concerned.

Currently, Clang does not support attaching comments to decls if there are preprocessor comments between the comment and the decl. This is enforced here: https://github.com/llvm/llvm-project/blob/b6ebea7972cd05a8e4dcf0d5a54f2c793999995a/clang/lib/AST/ASTContext.cpp#L284-L287

Alongside preprocessor directives, any instance of `;{}#@` between a comment and decl will cause the comment to not be attached to the decl.

#### Rationale

It would be nice for Clang-based documentation tools, such as [hdoc](https://hdoc.io), to support code using this pattern. Users expect to see comments attached to the relevant decl — even if there is an `#ifdef` in the way — which Clang does not currently do.

#### History

Originally, commas were also in the list of "banned" characters, but were removed in `b534d3a0ef69` ([link](https://github.com/llvm/llvm-project/commit/b534d3a0ef6970f5e42f10ba5cfcb562d8b184e1)) because availability macros often have commas in them. From my reading of the code, it appears that the original intent of the code was to exclude macros and decorators between comments and decls, possibly in an attempt to properly attribute comments to macros (discussed further in "Complications", below). There's some more discussion here: https://reviews.llvm.org/D125061.

### Change

This modifies Clang comment parsing so that comments are attached to subsequent declarations even if there are preprocessor directives between the end of the comment and the start of the decl. Furthermore, this change:

- Adds tests to verify that comments are attached to their associated decls even if there are preprocessor directives in between
- Adds tests to verify that current behavior has not changed (i.e. use of the other characters between comment and decl will result in the comment not being attached to the decl)
- Updates existing `lit` tests which would otherwise break.

#### Complications

Clang [does not yet support](https://github.com/llvm/llvm-project/issues/38206) attaching doc comments to macros. Consequently, the change proposed in this RFC affects cases where a doc comment attached to a macro is followed immediately by a normal declaration. In these cases, the macro's doc comments will be attached to the subsequent decl. Previously they would be ignored because any preprocessor directives between a comment and a decl would result in the comment not being attached to the decl. An example of this is shown below.

```cpp
/// Doc comment for a function-like macro
/// @param n
///    A macro argument
#define custom_sqrt(n) __internal_sqrt(n)

int __internal_sqrt(int n) { return __builtin_sqrt(n); }

// NB: the doc comment for the custom_sqrt macro will actually be attached to __internal_sqrt!
```

There is a real instance of this problem in the Clang codebase, namely here: https://github.com/llvm/llvm-project/blob/be10070f91b86a6f126d2451852242bfcb2cd366/clang/lib/Headers/amxcomplexintrin.h#L65-L114

As part of this RFC, I've added a semicolon to break up the Clang comment parsing so that the `-Wdocumentation` errors go away, but this is a hack. The real solution is to fix Clang comment parsing so that doc comments are properly attached to macros, however this would be a large change that is outside of the scope of this RFC.

### CODE REVIEW

**Comment by hdoc** [comment]
Ping

File: clang/lib/Headers/amxcomplexintrin.h [hunk 1]
```diff
@@ -62,6 +62,8 @@
 ///    The 2nd source tile. Max size is 1024 Bytes.
 #define _tile_cmmimfp16ps(dst, a, b) __builtin_ia32_tcmmimfp16ps(dst, a, b)
 
+;
```
[the remark this answers is not in the document]
Comment (hdoc) [reply clang/lib/Headers/amxcomplexintrin.h]: This has been fixed in the latest commit, where the internal function is documented with a Doxygen doc comment so that the `-Wdocumentation` warnings are not triggered.

**Comment by Boddlnagg** [comment]
Regarding the positioning of Doxygen comments, I just saw this and was wondering if it's possible to also support parsing of Doxygen comments that come after the signature, which happens to be the preferred style in our codebase ...

```c++
static int doSomething(int foobar)
/** \brief ...
 * \param foobar foo bar
 * \return  return value
 */
{
...
```

Doxygen accepts this, although I don't know if it's specified/documented as such.

**Comment by hdoc** [comment]
From my reading of the Doxygen documentation it seems like anything goes except putting the doc comment inside the actual body of a function. That being said, I think that change is out of the scope of this PR because it would likely require modifications to the Clang comment-to-decl matching functionality. It's better to do that in a separate PR

**Comment by AaronBallman** [comment]
Is this still in Draft status or should this be marked as ready for review?

**Comment by hdoc** [comment]
@AaronBallman this one is ready for review :)

File: clang/lib/Headers/amxcomplexintrin.h [hunk 2]
```diff
@@ -107,6 +107,24 @@
 ///    The 2nd source tile. Max size is 1024 Bytes.
 #define _tile_cmmrlfp16ps(dst, a, b) __builtin_ia32_tcmmrlfp16ps(dst, a, b)
 
+/// Perform matrix multiplication of two tiles containing complex elements and
```
Comment (AaronBallman) [inline clang/lib/Headers/amxcomplexintrin.h:110]: Is this an unrelated change?

File: clang/lib/Headers/ia32intrin.h [hunk 3]
```diff
@@ -533,6 +533,8 @@ __rdtscp(unsigned int *__A) {
 /// \see __rdpmc
 #define _rdpmc(A) __rdpmc(A)
 
+;
```
Comment (AaronBallman) [inline clang/lib/Headers/ia32intrin.h:536]: This is quite surprising; I'm guessing this is so that the previous comment does not associate with the function declared below, but this still feels pretty unclean.

**Review by AaronBallman — COMMENTED** [review]
Because doxygen supports documenting macros (https://www.doxygen.nl/manual/commands.html#cmddef), I am worried how often this will cause us to associate comments incorrectly on the declaration.

I wonder if we should be a bit smarter and check for `#define` at the start of a line when we encounter a `#`. e.g.,
```
/*!
  \def DERP(x)
  Does a derpy thing with x
*/
#define DERP(x) (x)

void derp(void); // Does not get the doxygen comment
```
while
```
/// Does amazing things, like works in the presence of
/// #define doing stupid things.
void func(); // does get the doxygen comment
```

File: clang/lib/Headers/amxcomplexintrin.h (hunk 2 above)
Comment (hdoc) [reply clang/lib/Headers/amxcomplexintrin.h:110]: This is not an unrelated change, I describe the issue in the "Complications" section of the PR body. In short, the `-Wdocumentation` warning gets triggered here because the macro for the `#define` on line 108 has a doc comment which Clang attributes to the function on line 128. Since the params documented in the macro's doc comment don't match the params in the function, `-Wdocumentation` throws an error. 

As a hack, I originally put semicolons on their own line because that effectively splits the decls from `-Wdocumentation`'s point of view, but it's a bit of a dirty hack. Instead, I just added a proper doc comment for the function here which stops the warning without changing the underlying comment parsing functionality. Still a workaround, but cleaner.

Changing that underlying functionality would be a much more involved since the comment parser operates on the AST, at which point most of the preproc info is gone because of how Clang is structured (please correct me if I'm wrong here).

File: clang/lib/Headers/ia32intrin.h (hunk 3 above)
Comment (hdoc) [reply clang/lib/Headers/ia32intrin.h:536]: If it sounds reasonable to you, I can fix this with the same method used in the `clang/lib/Headers/amxcomplexintrin.h` file. Unfortunately changing the underlying comment parsing functionality would be a complicated change.

**Review by AaronBallman — COMMENTED** [review]
Changes generally LG aside from an update to a test, but this should also have a release note in `clang/docs/ReleaseNotes.rst` so users know about the new functionality and the limitations regarding macros.

File: clang/lib/Headers/ia32intrin.h (hunk 3 above)
Comment (hdoc) [reply clang/lib/Headers/ia32intrin.h]: Agreed, I think it will need to change at some point and I will see if we can get support for doing that work.

**Comment by hdoc** [comment]
@AaronBallman Added the blurb to the release notes, let me know if that matches your expectations or if any changes are required. Thank you!

**Comment by hdoc** [comment]
Ping

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by hdoc** [comment]
This is rebased onto the latest rev of the main branch and should be ready to go. We do not have commit permissions, so @AaronBallman could you please help merge this in? Thank you :)

**Comment by gribozavr** [comment]
I'd like to push back against this change, I believe it is too broad. (Full disclosure: I'm the original author of this subsystem in Clang.)

We (Google) are integrating this change downstream and it causes a bunch of comments to get associated to unrelated declarations. The problem is particularly bad with file-level comments, which get attached to the first declaration.

For example: https://github.com/google/crubit/blob/main/rs_bindings_from_cc/test/golden/definition_of_forward_declaration.h

In this file the license comment gets attached to `ForwardDeclaredStruct` even though they are separated by an `#include` directive.

I think you should at the very minimum check for `#include` directives as a negative signal that breaks the connection.

**Comment by danakj** [comment]
`#include` and `#define` at a minimum should break the connection. Comments above a `#define` can be attributed to the macro defined there, which is something that Subdoc does (example result: https://suslib.cc/macro.sus_check.html)

**Comment by AaronBallman** [comment]
Thanks for the post-commit feedback! I've reverted the changes in 072e81db7a974bfb27b9b65d73330de7dd739821 so that @hdoc can investigate further.

**Comment by hdoc** [comment]
+1 on the revert and thank you for letting me know about this. I'll make changes to limit how aggressively the frontend attaches comments to decls inline with what @gribozavr and @danakj suggested.
